File: ios/Sources/MeasureSDK/Swift/Tracing/InternalSpan.swift
//
//  InternalSpan.swift
//  Measure
//
//  Created by Adwin Ross on 08/04/25.
//

import Foundation

/// Internal interface that extends `Span` with additional functionality needed by the SDK.
protocol InternalSpan: Span {
    /// Gets the name identifying this span.
    var name: String { get }

    /// Gets the session identifier associated with this span. A v4-UUID string.
    var sessionId: String { get }

    /// Gets the timestamp when this span was started.
    var startTime: Number { get }

    /// Gets the list of time-based checkpoints added to this span.
    var checkpoints: [Checkpoint] { get }

    /// Gets the map of attributes attached to this span.
    var attributes: Attributes? { get }

    /// Gets the current status of this span, indicating its outcome or error state.
    func getStatus() -> SpanStatus

    /// Returns a modifiable map of user-defined attributes.
    func getUserDefinedAttrs() -> [String: AttributeValue]

    /// Adds an attribute to this span.
    func setInternalAttribute(_ attribute: Attributes)

    /// Converts the span to a data class for further processing and export.
    func toSpanData() -> SpanData
}

File: react-native/src/tracing/internalSpan.ts
import type { ValidAttributeValue } from "../utils/attributeValueValidator";
import type { Checkpoint } from "./checkpoint";
import type { Span } from "./span";
import type { SpanData } from "./spanData";
import type { SpanStatus } from "./spanStatus";

export interface InternalSpan extends Span {
    /**
     * Gets the name identifying this span.
     */
    name: string;

    /**
     * Gets the timestamp when this span was started.
     */
    startTime: number;

    /**
     * Gets the list of time-based checkpoints added to this span.
     */
    checkpoints: Checkpoint[];

    /**
     * Gets the map of all attributes attached to this span (including internal ones).
     * Returns null if no attributes are set.
     */
    attributes: {[key: string]: any} | undefined;

    /**
     * Gets the current status of this span, indicating its outcome or error state.
     * @returns The current SpanStatus.
     */
    getStatus(): SpanStatus;

    /**
     * Returns a modifiable map (object) of user-defined attributes.
     * @returns A map of user-defined attributes.
     */
    getUserDefinedAttrs(): { [key: string]: ValidAttributeValue };

    /**
     * Adds an attribute intended for internal SDK use to this span.
     * @param attribute The attribute to set (often used for merging internal data).
     */
    setInternalAttribute(attribute: {[key: string]: any}): void;

    /**
     * Converts the span to a data class for further processing and export.
     * @returns A complete SpanData structure ready for serialization.
     */
    toSpanData(): SpanData;

    /**
     * Marks the span as sampled or not sampled.
     * @param sampled true if the span is selected for export, false otherwise.
     */
    setSampled(sampled: boolean): void;
}